File: scanner/app/src/main/java/com/claycode/scanner/topology_decoder/TextBitsConverter.kt
/*
 * Copyright (c) 2025 Claycode
 * Licensed under the MIT License with the Commons Clause.
 * Commercial use is prohibited without a separate license.
 * See LICENSE in the project root for details.
 * SPDX-License-Identifier: MIT AND Commons-Clause
 */

package com.claycode.scanner.topology_decoder

import com.claycode.scanner.data_structures.BitString

/**
 * This class provides conversions between text and bits.
 * Bits are represented as strings of zeros and ones.
 */
class TextBitsConverter {
    companion object {
        fun textToBits(text: String): BitString {
            val bytes = text.toByteArray(Charsets.UTF_8)
            val binaryStr = bytes.joinToString(separator = "") { byte ->
                String.format(
                    "%8s", byte.toInt()    // Convert to int
                        .and(255)    // keep last 8 bits only
                        .toString(2) // turn into binary digit
                ).replace(' ', '0') // Ensures leading zeros are maintained
            }
            return BitString(binaryStr)
        }

        fun bitsToText(bits: BitString): String {
            // Split the binary string into 8-bit segments that correspond to bytes
            val bytes = bits.toString().chunked(8).map { it.toInt(2).toByte() }.toByteArray()
            // Convert byte array back to string using UTF-8 encoding
            return String(bytes, Charsets.UTF_8)
        }
    }
}

File: webapp/common/bits/text_bits_converter.js
/*!
 * Copyright (c) 2025 Claycode
 * Licensed under the MIT License with the Commons Clause.
 * Commercial use is prohibited without a separate license.
 * See LICENSE in the project root for details.
 * SPDX-License-Identifier: MIT AND Commons-Clause
 */

import { BitString } from "./bit_string.js";
import { BitsValidator } from "./bits_validator.js";

/**
 * Common TextBitsConverter class used by both scanner and generator
 */
export class TextBitsConverter {
    static textEncoder = new TextEncoder("utf-8");
    static textDecoder = new TextDecoder("utf-8");
    
    static bitsToText(bits) {
        if (!bits) return "";
        
        // Convert array to string if needed
        let bitString;
        if (Array.isArray(bits)) {
            bitString = bits.join('');
        } else if (bits instanceof BitString) {
            bitString = bits.toString();
        } else if (typeof bits === 'string') {
            bitString = bits;
        } else {
            return "[Claycode] Invalid bit format";
        }
        
        try {
            // Split into 8-bit chunks and convert to bytes
            const chunks = bitString.match(/.{1,8}/g) || [];
            const bytes = [];
            
            for (const chunk of chunks) {
                if (chunk.length === 8) {
                    bytes.push(parseInt(chunk, 2));
                }
            }
            
            // Convert bytes to UTF-8 string
            const uint8Array = new Uint8Array(bytes);
            return this.textDecoder.decode(uint8Array);
        } catch (e) {
            return `[Claycode] ${bitString.length} bits`;
        }
    }

    static textToBits(text) {
        const encodedData = this.textEncoder.encode(text);
        const bitArray = this.uint8ArrayToBitArray(encodedData);
        return BitsValidator.addCRC(bitArray);
    }

    static uint8ArrayToBitArray(uint8Array) {
        const bitArray = [];
        for (let i = 0; i < uint8Array.length; i++) {
            const byte = uint8Array[i];
            for (let j = 7; j >= 0; j--) {
                bitArray.push((byte >> j) & 1);
            }
        }
        return bitArray;
    }

    static bitArrayToByteArray(bitArray) {
        const byteArray = [];
        for (let i = 0; i < bitArray.length; i += 8) {
            if (i + 8 <= bitArray.length) {
                let byte = 0;
                for (let j = 0; j < 8; j++) {
                    byte = (byte << 1) | bitArray[i + j];
                }
                byteArray.push(byte);
            }
        }
        return byteArray;
    }
}